### PR TITLE
WIP: Audio container

### PR DESCRIPTION
**Pull request by tomrf1**
## What does this change?
Supports the permanent /uk front container for the flagship podcast.
This is done by checking the name of the container + front.

This is the winner of the recent front A/B tests.

The container has 2 images:
1. The square image on the episode container is hardcoded in frontend
2. The image on the series snap card is taken from the fronts configuration

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [ ] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [ ] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [ ] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    allCards.collectFirst { case layout.FaciaCardAndIndex(_, card: layout.ContentCard, _) if card.snapStuff.isEmpty => card }
+}
+
+@getSnapLink(allCards: Seq[layout.FaciaCardAndIndex]) = @{
+    allCards.collectFirst { case layout.FaciaCardAndIndex(_, card: layout.ContentCard, _) if card.snapStuff.nonEmpty => card }
+}
+
+@subscriptionLinks = @{
+    <span>Subscribe</span>
+    <span><a data-link-name="apple" href="https://itunes.apple.com/gb/podcast/the-guardian-long-read/id587347784?mt=2">Apple Podcasts</a></span>
```
Comment (tomrf1) [inline common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: TODO - get the real subscription links

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        @for(slice <- containerLayout.slices.headOption) {
+
+            <div class="fc-slice-wrapper fc-podcast-container__main">
+                @defining(getAllCards(slice)) { allCards =>
+                    @getLatestEpisode(allCards).map { card =>
+                        <div class="fc-item--type-media">
+                            <div class="fc-podcast-container__episode">
+                                <div class="fc-podcast-container__episode-image">
+                                    @fragments.inlineSvg("audio-icon", "journalism/podcast-container", Seq("fc-podcast-container__audio-icon"))
+                                <img src="https://media.guim.co.uk/79284468f1b259db7d713dc24ea9af2a3f5c9937/0_0_800_800/500.png"/>
```
Comment (tomrf1) [inline common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: TODO - use the real image asset

File: common/app/views/fragments/containers/facia_cards/container.scala.html [hunk 3]
```diff
@@ -54,7 +57,15 @@
 
                     case _: Dynamic | _: Fixed => {
                         <div class="fc-container__inner">
-                            @standardContainer(containerDefinition, frontProperties, maybeContainerModel, showFrontBranding, frontId)
+                            @if(containerDefinition.displayName.contains("Flagship") && frontId.contains("uk")) {
```
Comment (tomrf1) [inline common/app/views/fragments/containers/facia_cards/container.scala.html]: TODO - use correct container name

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        @for(slice <- containerLayout.slices.headOption) {
+
+            <div class="fc-slice-wrapper fc-podcast-container__main">
+                @defining(getAllCards(slice)) { allCards =>
+                    @getLatestEpisode(allCards).map { card =>
+                        <div class="fc-item--type-media">
+                            <div class="fc-podcast-container__episode">
+                                <div class="fc-podcast-container__episode-image">
+                                    @fragments.inlineSvg("audio-icon", "journalism/podcast-container", Seq("fc-podcast-container__audio-icon"))
+                                    <img src="https://media.guim.co.uk/79284468f1b259db7d713dc24ea9af2a3f5c9937/0_0_800_800/500.png"/>
```
Comment (tomrf1) [inline common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: TODO - use the real image asset

File: common/app/conf/audio/Flagship.scala [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+  def isFlagshipContainer(id: String): Boolean =
+    FrontContainerIds.contains(id)
+
+  def displayFlagshipContainer(now: DateTime = DateTime.now): Boolean =
+    FlagshipFrontContainer.isSwitchedOn &&
+      now.isAfter(GoLiveDateTime) &&
+      now.getDayOfWeek != SATURDAY && now.getDayOfWeek != SUNDAY
+
+  val AlbumArtUrl = "https://media.guim.co.uk/79284468f1b259db7d713dc24ea9af2a3f5c9937/0_0_800_800/500.png"
```
Comment (tomrf1) [inline common/app/conf/audio/Flagship.scala]: TODO - update

File: static/src/stylesheets/module/facia-garnett/_container--podcast.scss [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+        >:first-child {
+            margin: 0 10px;
+            flex: calc(75% - 21px) 0 0;
+        }
+        >:last-child {
+            margin-left: 10px;
+            flex-basis: 25%;
+        }
+        >:last-child::before {
```
Comment (regiskuckaertz) [inline static/src/stylesheets/module/facia-garnett/_container--podcast.scss]: Well done on using the proper syntax for pseudo-elements 👍

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        column <- slice.columns
+        card <- column.cards
+    } yield card
+}
+
+@getLatestEpisode(allCards: Seq[layout.FaciaCardAndIndex]) = @{
+    allCards.collectFirst { case layout.FaciaCardAndIndex(_, card: layout.ContentCard, _) if card.snapStuff.isEmpty => card }
+}
+
+@getSnapLink(allCards: Seq[layout.FaciaCardAndIndex]) = @{
```
[the remark this answers is not in the document]
Comment (tomrf1) [reply common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: almost! `snapStuff.isEmpty` vs `snapStuff.nonEmpty`

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        @for(slice <- containerLayout.slices.headOption) {
+
+            <div class="fc-slice-wrapper fc-podcast-container__main">
+                @defining(getAllCards(slice)) { allCards =>
+                    @getLatestEpisode(allCards).map { card =>
+                        <div class="fc-item--type-media">
+                            <div class="fc-podcast-container__episode">
+                                <div class="fc-podcast-container__episode-image">
+                                    @fragments.inlineSvg("audio-icon", "journalism/podcast-container", Seq("fc-podcast-container__audio-icon"))
+                                    <img src="@FlagshipFrontContainer.AlbumArtUrl"/>
```
[the remark this answers is not in the document]
Comment (tomrf1) [reply common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: 👍

File: common/app/views/fragments/audio/containers/flagshipContainer.scala.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (regiskuckaertz) [reply common/app/views/fragments/audio/containers/flagshipContainer.scala.html]: I didn't see it 😵

File: common/app/conf/audio/Flagship.scala [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package conf.audio
+
+object Flagship {
```
Comment (tomrf1) [inline common/app/conf/audio/Flagship.scala]: this file was deleted in an earlier commit (replaced with FlagshipFrontContainer.scala)

**Comment by mattoh1** [comment]
💯

File: common/app/conf/audio/FlagshipFrontContainer.scala [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+object FlagshipFrontContainer {
+  private val FrontContainerIds = Seq(
+    "75ef80cd-2f3d-40d6-abf6-2021f88ece8e", //PROD
+    "c57a70c8-a00a-4a15-93a2-035b9221622b"  //CODE
+  )
+
+  private val GoLiveDateTime = DateTimeFormat.forPattern("yyyy/MM/dd HH:mm").parseDateTime(s"2018/11/01 03:15")
+
+  //The container should appear at 3:15 on Monday, and disappear at 3:15 on Saturday
```
Comment (regiskuckaertz) [inline common/app/conf/audio/FlagshipFrontContainer.scala]: Hmm, question: "3:15 on Saturday" should this be reflected in the email container PR?

Comment (tomrf1) [reply common/app/conf/audio/FlagshipFrontContainer.scala]: The requirements were different for the email

Comment (tomrf1) [reply common/app/conf/audio/FlagshipFrontContainer.scala]: (and now they're both the same!)

**Pull request merged**